File: 2- Elementary Sorts/Counting Sort/CountingSort.py
class CountingSort:
    @staticmethod
    def countSort(arr):  # for strings
        output = [0 for i in range(len(arr))]
        count = [0 for i in range(256)]
        ans = ["" for _ in arr]
        for i in arr:
            count[ord(i)] += 1
        for i in range(256):
            count[i] += count[i - 1]
        for i in range(len(arr)):
            output[count[ord(arr[i])] - 1] = arr[i]
            count[ord(arr[i])] -= 1
        for i in range(len(arr)):
            ans[i] = output[i]
        return ans

    @staticmethod
    def count_Sort(arr):  # for numbers
        max_element = int(max(arr))
        min_element = int(min(arr))
        range_of_elements = max_element - min_element + 1
        count_arr = [0 for _ in range(range_of_elements)]
        output_arr = [0 for _ in range(len(arr))]
        for i in range(0, len(arr)):
            count_arr[arr[i] - min_element] += 1
        for i in range(1, len(count_arr)):
            count_arr[i] += count_arr[i - 1]
        for i in range(len(arr) - 1, -1, -1):
            output_arr[count_arr[arr[i] - min_element] - 1] = arr[i]
            count_arr[arr[i] - min_element] -= 1
        for i in range(0, len(arr)):
            arr[i] = output_arr[i]
        return arr

File: 2- Elementary Sorts/Counting Sort/CountingSort.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <bits/stdc++.h>
#include <cstring>
using namespace std;
#define RANGE 255
class CountingSort {
public:
    static void countSort(char arr[]) {
        char output[strlen(arr)];
        int count[RANGE + 1], i;
        memset(count, 0, sizeof(count));
        for (i = 0; arr[i]; ++i)
            ++count[arr[i]];
        for (i = 1; i <= RANGE; ++i)
            count[i] += count[i - 1];
        for (i = 0; arr[i]; ++i) {
            output[count[arr[i]] - 1] = arr[i];
            --count[arr[i]];
        }
        for (i = 0; arr[i]; ++i)
            arr[i] = output[i];
    }
    static void countSort(vector<int>& arr){
        int max = *max_element(arr.begin(), arr.end());
        int min = *min_element(arr.begin(), arr.end());
        int range = max - min + 1;

        vector<int> count(range), output(arr.size());
        for (int i = 0; i < arr.size(); i++)
            count[arr[i] - min]++;

        for (int i = 1; i < count.size(); i++)
            count[i] += count[i - 1];

        for (int i = arr.size() - 1; i >= 0; i--) {
            output[count[arr[i] - min] - 1] = arr[i];
            count[arr[i] - min]--;
        }

        for (int i = 0; i < arr.size(); i++)
            arr[i] = output[i];
    }
};